### PR TITLE
Add support for ITypeParameterObjectCreationOperation to the CFG.

### PR DESCRIPTION
**Pull request by 333fred**
@dotnet/roslyn-compiler @AlekseyTs @dotnet/analyzer-ioperation for review.

### CODE REVIEW

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (jcouv) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: >operation.IsImplicit [](start = 179, length = 20)

In object creation we use `IsImplicit` (which also checks a flag for forcing implicit). Should it be used here too?

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (jcouv) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: >Return will [](start = 97, length = 11)

`Returns a reference...`?

**Review by jcouv — APPROVED** [review]
LGTM with one question. Thanks

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (333fred) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: Yes, good catch.

---
In reply to: [193148046](https://github.com/dotnet/roslyn/pull/27460#discussion_r193148046) [](ancestors = 193148046)

**Comment by 333fred** [comment]
@jcouv will wait for feedback from another review then address points.

**Review by jaredpar — APPROVED** [review]
:shipit:

**Comment by AlekseyTs** [comment]
>         public override IOperation VisitDynamicObjectCreation(IDynamicObjectCreationOperation operation, int? captureIdForResult)

Are you planning to handle this separately?

---
Refers to: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:5726 in 9fe34a1. [](commit_id = 9fe34a114ed9e06f3c1180e1f8537b2f5af9459e, deletion_comment = False)

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ITypeParameterObjectCreationTests.cs
Comment (mavasani) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ITypeParameterObjectCreationTests.cs]: >b ? 2 : 3 [](start = 26, length = 9)

We should probably have a control flow test with no control flow in the initializer.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ITypeParameterObjectCreationTests.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ITypeParameterObjectCreationTests.cs]: Consider adding a CFG test for this scenario

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 1). I am not signing off because you are planning to make some changes.

**Comment by 333fred** [comment]
>         public override IOperation VisitDynamicObjectCreation(IDynamicObjectCreationOperation operation, int? captureIdForResult)

Yes, this PR is just for ITypeParameterObjectCreationOperation.

---
In reply to: [394791716](https://github.com/dotnet/roslyn/pull/27460#issuecomment-394791716) [](ancestors = 394791716)

---
Refers to: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:5726 in 9fe34a1. [](commit_id = 9fe34a114ed9e06f3c1180e1f8537b2f5af9459e, deletion_comment = False)

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ITypeParameterObjectCreationTests.cs
Comment (333fred) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ITypeParameterObjectCreationTests.cs]: I didn't consider it relevant since it's not a type parameter object creation, this test was simply to confirm that behavior. I can add one to the invalidoperation tests though.

---
In reply to: [193162689](https://github.com/dotnet/roslyn/pull/27460#discussion_r193162689) [](ancestors = 193162689)

**Review by mavasani — APPROVED** [review]
:shipit:

**Review by AlekseyTs — APPROVED** [review]
LGTM (iteration 3)

**Comment by 333fred** [comment]
I ran tests locally, all IOperation tests passed. Merging to unblock fixing other conflicts.

**Pull request merged**
